Kernel/Term.cpp

```diff
@@ -678,14 +679,76 @@ std::string Term::toString(bool topLevel) const
     printArgs = isSort() || env.signature->getFunction(_functor)->combinator() == Signature::NOT_COMB;
   }
 
-  std::string s = headToString();
+#if NICE_THEORY_OUTPUT
+  auto theoryTerm = Kernel::tryNumTraits([&](auto numTraits) {
+    using NumTraits = decltype(numTraits);
+    auto uminus = [&]()  {
+    std::stringstream out;
+      auto maybePar = [&](auto t) { 
+        auto needsPar = t.isTerm() && NumTraits::isAdd(t.term()->functor());
+        return t.toString(!needsPar);
+      };
+      out << "-" << maybePar(termArg(0));
+      return Option<std::string>(out.str());
+    };
+    auto binary = [&](auto sym)  {
+      auto needsPar = !topLevel;
+      auto maybePar = [&](auto t) { 
+        return t.toString(
+            t.isTerm() && (
+              t.term()->functor() == _functor 
+              || (NumTraits::isAdd(_functor) && NumTraits::isMul(t.term()->functor()))
+              )
+            );
+      };
+      std::stringstream out;
+      out << (needsPar ? "(" : "");
+      out << maybePar(termArg(0)) << sym << maybePar(termArg(1));
+      out << (needsPar ? ")" : "");
+      return Option<std::string>(out.str());
+    };
+    if (isLiteral()) {
+      if (NumTraits::isGreater(_functor)) {
+        return binary(">");
+      } else if (NumTraits::isGeq(_functor)) {
+        return binary(">=");
+      }
+      /* nothing */
+    } else if (isSort()) {
+      /* nothing */
+    } else {
+      if (NumTraits::isAdd(_functor)) {
+        return binary(" + ");
+      } else if (NumTraits::isMul(_functor)) {
+        return binary(" ");
+      } else if (NumTraits::isFloor(_functor)) {
+        return some(outputToString("⌊", termArg(0), "⌋"));
+      } else if (NumTraits::isMinus(_functor)) {
+        return uminus();
+      }
+    }
+    return Option<std::string>();
```

Comment:
When does the "none" case occur? Could also consider using the empty string to signal that if the other cases are guaranteed non-empty.

Reply:
Added a clarifing comment. In this case we got an uninterprete term. Then we just format as usual.